homeworks/js-basics-1/solutions.js

```diff
@@ -0,0 +1,367 @@
+        case '+':
+            return value1 + value2;
+        case '-':
+            return value1 - value2;
+        case '*':
+            return value1 * value2;
+        case '/':
+            return value1 / value2;
+        default:
+            return NaN;
+    }
+}
+
+// http://www.codewars.com/kata/printing-array-elements-with-comma-delimiters
+function printArray(array) {
+    return array.join(',');
+}
+
+// http://www.codewars.com/kata/transportation-on-vacation
+function rentalCarCost(d) {
+    let discount = 0;
+    if (d >= 7) {
+        discount = 50;
+    } else if (d >= 3) {
+        discount = 20;
+    }
+    const costPerDay = 40;
+    return costPerDay * d - discount;
+}
+
+// http://www.codewars.com/kata/calculating-with-functions
+function zero(f) {
+    return f ? f(0) : 0;
+}
+function one(f) {
+    return f ? f(1) : 1;
+}
+function two(f) {
+    return f ? f(2) : 2;
+}
+function three(f) {
+    return f ? f(3) : 3;
+}
+function four(f) {
+    return f ? f(4) : 4;
+}
+function five(f) {
+    return f ? f(5) : 5;
+}
+function six(f) {
+    return f ? f(6) : 6;
+}
+function seven(f) {
+    return f ? f(7) : 7;
+}
+function eight(f) {
+    return f ? f(8) : 8;
+}
+function nine(f) {
+    return f ? f(9) : 9;
+}
+
+function plus(a) {
+    return (b) => b + a;
+}
+function minus(a) {
+    return (b) => b - a;
+}
+function times(a) {
+    return (b) => b * a;
+}
+function dividedBy(a) {
+    return (b) => Math.floor(b / a);
+}
+
+// http://www.codewars.com/kata/get-the-middle-character
+function getMiddle(s) {
+    const mid = Math.floor(s.length / 2);
+    if (s.length % 2 === 1) {
+        return s[mid];
+    }
+    return s.slice(mid - 1, mid + 1);
+}
+
+// http://www.codewars.com/kata/partition-on
+function partitionOn(pred, items) {
+    let left = [];
+    let right = [];
+    for (let item of items) {
+        if (!pred(item)) {
+            left.push(item);
+        } else {
+            right.push(item);
+        }
+    }
+    items.length = 0;
+    items.push(...left, ...right);
+    return left.length;
+}
+
+// https://www.codewars.com/kata/526571aae218b8ee490006f4
+function countBits(n) {
+    let onesCount = 0;
+    let reminder;
+    do {
+        reminder = n % 2;
+        if (reminder === 1) {
+            onesCount++;
+        }
+        n = Math.floor(n / 2);
+    } while (n > 0);
+    return onesCount;
+}
+
+// https://www.codewars.com/kata/find-the-odd-int/
+function findOdd(A) {
+    let count = new Map();
+    for (let item of A) {
+        if (!count.has(item)) {
+            count.set(item, 1);
+        } else {
+            count.set(item, count.get(item) + 1);
+        }
+    }
+    for (let key of count.keys()) {
+        if (count.get(key) % 2 === 1) {
+            return key;
+        }
+    }
+    return null;
+}
+
+// https://www.codewars.com/kata/find-the-parity-outlier
+function findOutlier(integers) {
+    let evenCount = integers.reduce((acc, int) => acc + (Math.abs(int % 2) === 0), 0)
```

Comment:
Task warns that, array of numbers could be very large. Thata means looping through it will take performance and time. The point was to break the cycle in the level of distinguished number and to save performance. Iterate through huge array, if it is possible to avoid, is a very bad idea.